# 极客时间 - 程序员基础数学课 - [第41课 - 线性回归(下)](https://time.geekbang.org/column/article/86766)



先不考虑截距的情况下计算：

In [1]:
from numpy import *

x = mat([[0,1],[1,-1],[2,8]])
y = mat([[1.4],[-0.48],[13.2]])

# 分别求出矩阵X'、X'X、(X'X)的逆
# 注意，这里的I表示逆矩阵而不是单位矩阵
print("X矩阵的转置X'：\n", x.transpose())
print("\nX'点乘X：\n", x.transpose().dot(x))
print("\nX'X矩阵的逆\n", (x.transpose().dot(x)).I)

print("\nX'X矩阵的逆点乘X'\n", (x.transpose().dot(x)).I.dot(x.transpose()))
print("\n系数矩阵B：\n", (x.transpose().dot(x)).I.dot(x.transpose()).dot(y))

X矩阵的转置X'：
 [[ 0  1  2]
 [ 1 -1  8]]

X'点乘X：
 [[ 5 15]
 [15 66]]

X'X矩阵的逆
 [[ 0.62857143 -0.14285714]
 [-0.14285714  0.04761905]]

X'X矩阵的逆点乘X'
 [[-0.14285714  0.77142857  0.11428571]
 [ 0.04761905 -0.19047619  0.0952381 ]]

系数矩阵B：
 [[0.93828571]
 [1.4152381 ]]


使用 sklearn 库中的 LinearRegression().fit() 函数 

In [1]:
import pandas as pd
from sklearn.linear_model import LinearRegression

df = pd.read_csv("linear-regression.csv")
df_features = df.drop(['y'], axis=1)     #Dataframe中除了最后一列，其余列都是特征，或者说自变量
df_targets = df['y']                     #Dataframe最后一列是目标变量，或者说因变量

print(df_features, df_targets)
#使用特征和目标数据，拟合线性回归模型
regression = LinearRegression().fit(df_features, df_targets)
#拟合程度的好坏
print(regression.score(df_features, df_targets))
#截距
print(regression.intercept_)
#各个特征所对应的系数
print(regression.coef_)

   x1  x2
0   0   1
1   1  -1
2   2   8 0     1.40
1    -0.48
2    13.20
Name: y, dtype: float64
1.0
-0.014545454545452863
[0.94909091 1.41454545]


在最终的结果中，1.0 表示拟合程度非常好，而 -0.014545454545452863 表示一个截距，[0.94909091 1.41454545]表示系数 b1​ 和 b2​ 的值。这个结果和我们最小二乘法的结果有所差别，主要原因是 LinearRegression().fit() 默认考虑了有线性函数存在截距的情况。那么我们使用最小二乘法是不是也可以考虑有截距的情况呢？答案是肯定的，不过我们首先要略微修改一下方程组和矩阵 X。如果我们假设有截距存在，那么线性回归方程就要改写为：

In [2]:
from numpy import *

x = mat([[1,0,1],[1,1,-1],[1,2,8]])
y = mat([[1.4],[-0.48],[13.2]])

print("\n系数矩阵B：\n", (x.transpose().dot(x)).I.dot(x.transpose()).dot(y))


系数矩阵B：
 [[-0.01454545]
 [ 0.94909091]
 [ 1.41454545]]


这个结果和 LinearRegression().fit() 的结果就一致了。